planner_cspace/test/src/test_grid_astar.cpp

```diff
@@ -0,0 +1,172 @@
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <list>
+#include <vector>
+
+#include <boost/thread.hpp>
+
+#include <omp.h>
+
+#include <gtest/gtest.h>
+
+#include <planner_cspace/grid_astar.h>
+
+TEST(GridAstar, ParallelSearch)
+{
+  using Vec = CyclicVecInt<1, 1>;
+  GridAstar<1, 1> as(Vec(16));
+  as.setSearchTaskNum(8);
+  omp_set_num_threads(2);
+
+  const auto cb_cost = [](
+      const Vec&, Vec&, const Vec&, const Vec&) -> float
+  {
+    return 1.0;
+  };
+  const auto cb_cost_estim = [](const Vec&, const Vec&) -> float
+  {
+    return 1e-6;
+  };
+  std::vector<Vec> search[16];
+  for (int i = 1; i < 15; ++i)
+  {
+    search[0].push_back(Vec(i));
+    for (int j = 1; j < 15; ++j)
+    {
+      if (i == j)
+        continue;
+      search[i].push_back(Vec(j));
+    }
+    search[i].push_back(Vec(15));
+  }
+
+  const auto cb_search = [&search](
+      const Vec& p, const Vec&, const Vec&) -> std::vector<Vec>&
+  {
+    return search[p[0]];
+  };
+  const auto cb_progress = [](const std::list<Vec>&)
+  {
+    return true;
+  };
+
+  for (int i = 0; i < 1000; ++i)
+  {
+    std::list<Vec> path;
+    ASSERT_TRUE(
+        as.search(
+            Vec(0), Vec(15), path,
+            cb_cost, cb_cost_estim, cb_search, cb_progress,
+            0, 1.0));
+    ASSERT_EQ(path.size(), 3u);
+    ASSERT_EQ(path.front(), Vec(0));
+    ASSERT_EQ(path.back(), Vec(15));
+  }
+}
+
+class GridAstarTestWrapper : public GridAstar<1, 1>
+{
+public:
+  explicit GridAstarTestWrapper(const Vec& size)
+    : GridAstar(size)
+  {
+  }
+  std::unordered_map<Vec, Vec, Vec>& parentMap()
+  {
+    return parents_;
+  }
+  bool findPath(const Vec& s, const Vec& e, std::list<Vec>& path)
+  {
+    return GridAstar::findPath(s, e, path);
+  }
+};
+
+TEST(GridAstar, FindPathLooped)
+{
+  using Vec = GridAstarTestWrapper::Vec;
+  GridAstarTestWrapper as(Vec(4));
+
+  as.parentMap()[Vec(3)] = Vec(2);
+  as.parentMap()[Vec(2)] = Vec(1);
```

Comment:
Similar endless loop could be caused by the very much rare condition in parallel search.
It might caused if two jobs calculate cost to each other within (maybe) a few nano seconds and these grids are not updated afterward. Especially, two jobs running in the different CPU core might increases probability due to cache update duration.